File: core/state.ts
// core/state.ts — SessionState, FSM, Memento, Command
import { randomUUID } from 'node:crypto'
import {
  AgentFSMState,
  Command,
  FSMTrace,
  FSMViolationError,
  ISessionState,
  Memento,
  StateObserver,
  Unsubscribe,
} from './protocols.js'

// Valid FSM transitions per CLAUDE.md §12
const VALID_TRANSITIONS = new Map<AgentFSMState, Set<AgentFSMState>>([
  ['IDLE',       new Set<AgentFSMState>(['GATHERING'])],
  ['GATHERING',  new Set<AgentFSMState>(['MODELING', 'AWAITING', 'FAILURE'])],
  ['AWAITING',   new Set<AgentFSMState>(['MODELING', 'FAILURE'])],
  ['MODELING',   new Set<AgentFSMState>(['SOLVING', 'FAILURE'])],
  ['SOLVING',    new Set<AgentFSMState>(['VALIDATING', 'FAILURE'])],
  ['VALIDATING', new Set<AgentFSMState>(['DELIVERING', 'FAILURE'])],
  ['DELIVERING', new Set<AgentFSMState>(['IDLE'])],
  ['FAILURE',    new Set<AgentFSMState>(['DELIVERING'])],
  ['DEGRADED',   new Set<AgentFSMState>(['IDLE', 'FAILURE'])],
])

export class SessionState implements ISessionState {
  private _fsm: AgentFSMState = 'IDLE'
  private _turn = 0
  private _contextName = 'NullContext'
  private _scratch: Record<string, unknown> = {}
  private _trace: FSMTrace = []
  private readonly _mementos = new Map<string, Memento>()
  private readonly _commandLog: Command[] = []
  private readonly _observers = new Set<StateObserver>()
  private readonly _maxMementos: number

  constructor(maxMementos = 100) {
    this._maxMementos = maxMementos
  }

  get fsm(): AgentFSMState { return this._fsm }
  get turn(): number { return this._turn }
  get contextName(): string { return this._contextName }
  get trace(): FSMTrace { return [...this._trace] }
  get history(): Command[] { return [...this._commandLog] }

  transition(to: AgentFSMState): void {
    const allowed = VALID_TRANSITIONS.get(this._fsm)
    if (!allowed?.has(to)) {
      throw new FSMViolationError(this._fsm, to, this.trace)
    }
    this._fsm = to
    this._trace.push({ state: to, timestamp: Date.now() })
    this._notifyObservers()
  }

  setContext(name: string): void {
    this._contextName = name
  }

  snapshot(): Memento {
    if (this._mementos.size >= this._maxMementos) {
      // Evict oldest (LRU)
      const oldestKey = this._mementos.keys().next().value
      if (oldestKey !== undefined) {
        this._mementos.delete(oldestKey)
      }
    }
    const memento: Memento = {
      id: randomUUID(),
      fsm: this._fsm,
      turn: this._turn,
      contextName: this._contextName,
      scratch: { ...this._scratch },
      timestamp: Date.now(),
    }
    this._mementos.set(memento.id, memento)
    return memento
  }

  restore(id: string): void {
    const memento = this._mementos.get(id)
    if (!memento) throw new Error(`Memento not found: ${id}`)
    this._fsm = memento.fsm
    this._contextName = memento.contextName
    this._scratch = { ...memento.scratch }
    this._trace = []
    this._notifyObservers()
  }

  commit(cmd: Command): void {
    this._commandLog.push(cmd)
    this._turn += 1
    this._notifyObservers()
  }

  subscribe(obs: StateObserver): Unsubscribe {
    this._observers.add(obs)
    return () => { this._observers.delete(obs) }
  }

  async flush(): Promise<void> {
    // Base: no-op. Override for persistence.
  }

  reset(): void {
    this._fsm = 'IDLE'
    this._trace = []
    this._scratch = {}
    this._notifyObservers()
  }

  private _notifyObservers(): void {
    for (const obs of this._observers) {
      obs(this._fsm)
    }
  }
}

File: src/ooagent/core/state.py
"""core/state.py — SessionState, FSM, Memento, Command."""

from __future__ import annotations

import time
import uuid
from collections import OrderedDict

from ooagent.core.protocols import (
    AgentFSMState,
    Command,
    FSMTrace,
    FSMTraceEntry,
    FSMViolationError,
    ISessionState,
    Memento,
    StateObserver,
    Unsubscribe,
)

# Valid FSM transitions per CLAUDE.md §12
VALID_TRANSITIONS: dict[AgentFSMState, set[AgentFSMState]] = {
    "IDLE": {"GATHERING"},
    "GATHERING": {"MODELING", "AWAITING", "FAILURE"},
    "AWAITING": {"MODELING", "FAILURE"},
    "MODELING": {"SOLVING", "FAILURE"},
    "SOLVING": {"VALIDATING", "FAILURE"},
    "VALIDATING": {"DELIVERING", "FAILURE"},
    "DELIVERING": {"IDLE"},
    "FAILURE": {"DELIVERING"},
}


class SessionState(ISessionState):
    def __init__(self, max_mementos: int = 100) -> None:
        self._fsm: AgentFSMState = "IDLE"
        self._turn = 0
        self._context_name = "NullContext"
        self._scratch: dict[str, object] = {}
        self._trace: FSMTrace = []
        self._mementos: OrderedDict[str, Memento] = OrderedDict()
        self._command_log: list[Command] = []
        self._observers: set[StateObserver] = set()
        self._max_mementos = max_mementos

    @property
    def fsm(self) -> AgentFSMState:
        return self._fsm

    @property
    def turn(self) -> int:
        return self._turn

    @property
    def context_name(self) -> str:
        return self._context_name

    @property
    def trace(self) -> FSMTrace:
        return list(self._trace)

    @property
    def history(self) -> list[Command]:
        return list(self._command_log)

    def transition(self, to: AgentFSMState) -> None:
        allowed = VALID_TRANSITIONS.get(self._fsm, set())
        if to not in allowed:
            raise FSMViolationError(self._fsm, to, self.trace)
        self._fsm = to
        self._trace.append(FSMTraceEntry(state=to, timestamp=time.time()))
        self._notify_observers()

    def set_context(self, name: str) -> None:
        self._context_name = name

    def snapshot(self) -> Memento:
        if len(self._mementos) >= self._max_mementos:
            self._mementos.popitem(last=False)  # evict least-recently-used
        memento = Memento(
            id=str(uuid.uuid4()),
            fsm=self._fsm,
            turn=self._turn,
            context_name=self._context_name,
            scratch=dict(self._scratch),
            timestamp=time.time(),
        )
        self._mementos[memento.id] = memento
        return memento

    def restore(self, id: str) -> None:
        memento = self._mementos.get(id)
        if memento is None:
            raise ValueError(f"Memento not found: {id}")
        self._mementos.move_to_end(id)  # mark as most-recently-used
        self._fsm = memento.fsm
        self._turn = memento.turn
        self._context_name = memento.context_name
        self._scratch = dict(memento.scratch)
        self._trace = []
        self._notify_observers()

    def commit(self, cmd: Command) -> None:
        self._command_log.append(cmd)
        self._turn += 1
        self._notify_observers()

    def subscribe(self, obs: StateObserver) -> Unsubscribe:
        self._observers.add(obs)

        def unsubscribe() -> None:
            self._observers.discard(obs)

        return unsubscribe

    async def flush(self) -> None:
        """Base: no-op. Override for persistence."""

    def reset(self) -> None:
        self._fsm = "IDLE"
        self._trace = []
        self._scratch = {}
        self._notify_observers()

    def _notify_observers(self) -> None:
        for obs in self._observers:
            obs(self._fsm)
